src/FrameBase.re

```diff
@@ -60,30 +58,287 @@ external waitForXPath :
   Js.Promise.t(ElementHandle.t) =
   "";
 
-/* TODO: Add support [, ...args] */
+/**
+ * selectOneEval(selector, fn, page)
+ * Runs document.querySelector in the page and passes it as the first
+ * argument to `fn`. If there's no element matching selector, the function
+ * throws an error.
+ */
+[@bs.send.pipe : t]
+external selectOneEval : (string, Dom.element => 'r) => Js.Promise.t('r) =
+  "$eval";
+
+/**
+ * selectOneEvalPromise(selector, fn, page)
+ * Runs document.querySelector in the page and passes it as the first
+ * argument to `fn`. If there's no element matching selector, the function
+ * throws an error. `fn` must return a promise.
+ */
+[@bs.send.pipe : t]
+external selectOneEvalPromise :
+  (string, Dom.element => Js.Promise.t('r)) => Js.Promise.t('r) =
```

Comment:
Just curious, why need `Promise` version (`Dom.element => Js.Promise.t('r)`) here? Would `Dom.element => 'r` not be enough?

Reply:
Then the return type of the external is potentially `Js.Promise.t(Js.Promise.t('r))` or `'r` which are both incorrect. Javascript promises can't be nested like that and are flattened to one under the hood.. The puppeteer docs also state it will always only return one promise. We need the return type from inside the promise returned from the evaluation function, otherwise we can indicated a return type that is not a promise at all if they pass a function that does not return a promise. I know it seems odd but I'm not sure that there are any other straightforward ways of dealing with this.

Reply:
If you need to eval a function that does not return a promise then use `selectOneEval` instead of this function. The types will keep you honest as long as there is a known return type (i.e. not `'a` or the like). On a side note, we're going to want to write and generate API docs at some point since we are deviating a bit from the upstream API, and it's nice to see it in the editor. I've started to write some here.